vendor/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_image.go

```diff
@@ -31,8 +31,7 @@ import (
 // secrets if necessary.
 func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (string, error) {
 	img := image.Image
-	repoToPull, _, _, err := parsers.ParseImageName(img)
-	if err != nil {
+	if _, _, _, err := parsers.ParseImageName(img); err != nil {
```

Comment:
recording here for future action.

we should work to change upstream to not depend on `dockerref.ParseNormalizedNamed` in favor of a utility that makes no assumptions about unofficial image name standards.

fyi @mrunalp @runcom - it would be great if we have a corresponding library in containers/image that we can use in favor of dockerref in areas in front of CRI in upstream that does the same function...

Reply:
c/image uses the same `ParseNormalizedNamed` ~throughout (but some of its users also support unqualified names).

More and more use cases for a non-normalizing parser (either accepting arbitrary [host/]repo without modification, or requiring an explicit host name) keep popping up, so it may well make sense for c/image to provide a parser for that — but that would probably need to be be very explicitly an incompatible type, to avoid mixing the unqualified/non-normalized/normalized references unintentionally.

Reply:
(The code as is, parsing the input _only_ for syntax validity and ignoring the results of the parsing, can just use `ParseNormalizedNamed`, equivalently from docker/distribution/reference or c/image/docker/reference ; of course there’s a risk that someone will re-add uses of the normalized value and regress.)

Reply:
i would like to have a neutral library that makes no qualification that assumes anything about a particular vendor, and do the leg work to make it popular in impacted projects.  using the results makes it have inadvertent side effects that change user input.

Reply:
> i would like to have a neutral library that makes no qualification that assumes anything about a particular vendor, and do the leg work to make it popular in impacted projects. 

What would the library do about names like `busybox`, though? Just refusing them would be massively disruptive (though that is easy already - `github.com/{docker/distribution,containers/image/docker}/reference.ParseNamed` does that); if they are accepted, they have to be interpreted _somehow_.

It’s not just a matter of not making vendor preferences about hostnames, dealing with hostname-less names is a non-trivial design problem (can hostname-less names like `busybox` exist stand-alone? If not, what happens when someone tags an image with such a name? If they can, what happens when someone specifies a `busybox` image in a container specification - will it try to pull something, and from where? Will it prefer a local `busybox` image if it somehow got tagged, or ignore it prefer to pull a remote one?

Allowing hostname-less names ends up with not being a “name parsing” problem, but a library to do all pull/push/tag/lookup , interacting _both_ with remote registries and local storage.

Something like that may well be created between libpod/buildah/CRI-O , but to be clear, the end result from k8s side will very likely continue to be that 1) image names should be treated as opaque objects not to be parsed or touched if at all possible, and 2) users will almost certainly want to continue to explicitly specify hostnames in container specifications (e.g. `podman tag busybox` currently names it `localhost/busybox:latest`, seems not too useful in a cluster.)